helper.py

```diff
@@ -112,13 +112,29 @@ def display_ui(self, entity, enemies, combat=True):
             ansiprint(str(entity))
         print()
 
-    def list_input(self, input_string: str, array: list) -> int | None:
-        try:
-            ansiprint(input_string, end='')
-            option = int(input()) - 1
-            array[option] = array[option] # Checks that the number is in range but doesn't really do anything
-        except (ValueError, IndexError):
-            return None
+    def list_input(self, input_string: str, choices: list, conditions: tuple[str]=None) -> int | None:
+        if conditions is None:
+            conditions = ()
+        conditions.append("Valid") # Default condition that is always considered.
+        while True:
+            try:
+                ansiprint(input_string, end='')
+                option = int(input()) - 1
+                parameters: dict = {"Valid": option in range(0, len(choices)), "Upgradeable": choices[option].get("Upgraded") is False and choices[option].get("Effects+"),
+                                    "Removable": choices[option].get("Removable") is False, "Upgraded": choices[option].get("Upgraded") is True}
+                if not all((parameters[condition] for condition in conditions)):
+                    rules_broken = (condition.lower() for condition in conditions if parameters[condition] is False)
+                    list_of_rules_broken = ", ".join(rules_broken)
+                    ansiprint("\u001b[1A\u001b[1000D<red>The choice you selected is not " + list_of_rules_broken[:list_of_rules_broken.rindex(", ") + 2] + "or " + list_of_rules_broken[list_of_rules_broken.rindex(", ") + 2:] + "</red>", end='')
+                    sleep(1.5)
+                    print("\u001b[2K")
+                    continue
+            except ValueError:
```

Comment:
Not catching out of bounds index

Reply:
The out of bounds check was the "Valid" Key in the dictionary. I wanted to have the key in the set of conditions by default.

Reply:
That creates a boolean, but doesn't stop you from using it.

Ex: Pretend option is something huge like 999.

```python
option = int(input()) - 1    # 999
parameters: dict = {    # these aren't lambdas or partials -- this entire dictionary is evaluated on the spot
   "Valid": option in range(0, len(choices)),     # evaluates to False
   "Upgradeable": choices[option].get("Upgraded") is False and choices[option].get("Effects+"),  # what happens here with choices[999] ?
   "Removable": choices[option].get("Removable") is False,   # what happens here with choices[999] ?
   "Upgraded": choices[option].get("Upgraded") is True    # what happens here with choices[999] ?
}
```

You can code up a simple test too, something like:

```python
def test_list_input():
  all_cards = list(items.cards.values())
  result = helper.view.list_input(input_string='Choose a card: ', choices=all_cards)
  print(f"You got {result}")
```

and run it with `pytest --capture=no` so that it'll let you interact with it, and try a number like 999.

Reply:
Yea, Ima do the function validator stuff.